server/service/src/print/jetdirect.rs

```diff
@@ -0,0 +1,89 @@
+use anyhow::Result;
+use std::io::Write;
+use std::time::Duration;
+use telnet::{Event, Telnet};
+
+pub struct Jetdirect {
+    pub addr: String,
+    port: u16,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum Mode {
+    Sgd,
+    Print,
+}
+
+impl Jetdirect {
+    pub fn new(addr: String, port: u16) -> Jetdirect {
+        Jetdirect { addr, port }
+    }
+}
+
+impl Jetdirect {
+    fn send_command_and_print(
+        &self,
+        payload: String,
+        handle: &mut telnet::Telnet,
+        mode: Mode,
+        timeout: Duration,
+    ) -> Result<()> {
+        handle.write(payload.as_bytes())?;
+        // As far as I can tell, there's no way to detect the end of an SGD command response.
+        // There can be any number of double-quotes; there's no terminating control character, newline, etc.
+        // Only thing we can really do is print lines as we get them, and wait for a timeout.
+        let mut resp = String::new();
+        loop {
+            let event = handle.read_timeout(timeout)?;
+            match event {
+                Event::Data(data) => {
+                    if mode == Mode::Sgd {
+                        let resp_part = String::from_utf8_lossy(&data);
+                        resp.push_str(&resp_part);
+                        std::io::stdout().flush()?;
+                        if return_early(&data) {
+                            break;
+                        }
+                    }
+                }
+                Event::TimedOut => {
+                    // We don't get the line break at the end of a response, usually
+                    break;
+                }
+                _ => {
+                    println!("Got other jetdirect event: {:?}", event)
+                }
+            }
+        }
+        if !resp.is_empty() {
+            println!("{}", resp);
+        }
+
+        Ok(())
+    }
+
+    pub fn send_string(&self, data: String, mode: Mode) -> Result<()> {
+        let mut telnet = Telnet::connect((self.addr.clone(), self.port), 512)?;
```

Comment:
```suggestion
        let mut telnet = Telnet::connect_timeout((self.addr.clone(), self.port), 512)?;
```

You actually need to do something like 

SocketAddr::new(IpAddr::V4(Ipv4Addr::from_str(self.addr.clone())
                                .expect("Invalid address")), self.port)

But that could be done once and stored in the `Jetdirect` struct...

This isn't essential as it does seem to timeout after 60s for me anyway